contractsV5/KyberNetwork.sol

```diff
@@ -422,62 +423,61 @@ contract KyberNetwork is Withdrawable, Utils, IKyberNetwork, ReentrancyGuard {
     // Regarding complexity. Below code follows the required algorithm for choosing a reserve.
     //  It has been tested, reviewed and found to be clear enough.
     //@dev this function always src or dest are ether. can't do token to token
-    function searchBestDestAmount(address[] memory reserveArr, IERC20 src, IERC20 dest, uint srcAmount, uint takerFeeBps)
+    function searchBestRate(IKyberReserve[] memory reserveArr, IERC20 src, IERC20 dest, uint srcAmount, uint takerFeeBps)
         public
         view
-        returns(address reserve, uint destAmount, bool isPayingFees)
+        returns(IKyberReserve reserve, uint bestRate, bool isPayingFees)
     {
-        //todo: read isPayingFees once.
+        //use destAmounts for comparison, but return the best rate
         uint bestDestAmount = 0;
         uint bestReserve = 0;
-        uint numRelevantReserves = 1; // assume alywas best reserve will be relevant
+        uint numRelevantReserves = 1; // assume always best reserve will be relevant
 
         //return 1 for ether to ether
-        if (src == dest) return (address(reserves[bestReserve]), PRECISION, false);
+        if (src == dest) return (reserves[bestReserve], PRECISION, false);
 
-        if (reserveArr.length == 0) return (address(reserves[bestReserve]), 0, false);
+        if (reserveArr.length == 0) return (reserves[bestReserve], 0, false);
 
-        uint[] memory destAmounts = new uint[](reserveArr.length);
+        uint[] memory rates = new uint[](reserveArr.length);
         uint[] memory reserveCandidates = new uint[](reserveArr.length);
-        uint i;
-        uint rate;
-        uint srcAmountWithFee;
-
-        for (i = 0; i < reserveArr.length; i++) {
-            //list all reserves that have this token.
-
-            srcAmountWithFee = (src == ETH_TOKEN_ADDRESS && isFeeLessReserve[reserveArr[i]])? srcAmount : 
-                srcAmount * (BPS - takerFeeBps) / BPS;
-            rate = (IKyberReserve(reserveArr[i])).getConversionRate(
-                src, 
-                dest, 
-                srcAmountWithFee, 
+        uint destAmountForComparison;
+        uint fullDestAmount;
+
+        for (uint i = 0; i < reserveArr.length; i++) {
+            //list all reserves that support this token.
+            isPayingFees = isFeePayingReserve[address(reserveArr[i])]; //not feeless
+            rates[i] = reserveArr[i].getConversionRate(
+                src,
+                dest,
+                (src == ETH_TOKEN_ADDRESS && isPayingFees) ? srcAmount * (BPS - takerFeeBps) / BPS : srcAmount, //srcAmountWithFee
                 block.number);
 
-            destAmounts[i] = srcAmountWithFee * rate / PRECISION;
-            destAmounts[i] = (dest == ETH_TOKEN_ADDRESS && isFeeLessReserve[reserveArr[i]])? destAmounts[i] : 
-                destAmounts[i] * (BPS - takerFeeBps) / BPS;
-            
-            if (destAmounts[i] > bestDestAmount) {
+            fullDestAmount = srcAmount * rates[i] / PRECISION;
```

Comment:
must user getDestAmount with decimals)

its my fault. I did it. :(
anyway better if u fix it in your PR

Reply:
Errmmmm it means I will need the decimals.....